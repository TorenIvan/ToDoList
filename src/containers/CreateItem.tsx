import { Platform, StyleSheet, TextInput, View, Keyboard } from "react-native";
import { RadioButtonValueType, itemHeight, Strings } from "../utils";
import React, { useState, useRef, useEffect } from "react";
import { useTheme } from "../store/globalTheme";
import RadioButton from "../components/RadioButton";

interface Props {
  onSubmit: ({ checked, text }: { checked: RadioButtonValueType; text: String }) => void;
}

const CreateItem: React.FC<Props> = ({ onSubmit }): JSX.Element => {
  const [checked, setChecked] = useState<RadioButtonValueType>(Strings.Unchecked);
  const [newText, setNewText] = useState<String>(Strings.EmptyString);
  const { theme } = useTheme();

  const inputRef = useRef<TextInput>(null);

  useEffect(() => {
    const keyboardDidHideListener = Keyboard.addListener(
      'keyboardDidHide',
      () => {
        handleBackButtonClick();
      }
    );
    return () => {
      keyboardDidHideListener.remove();
    };
  }, []);

  const handleBackButtonClick: any = () => {
    inputRef?.current?.blur();
  };

  const handleRadioButtonPress = () => {
    setChecked(checked == Strings.Checked ? Strings.Unchecked : Strings.Checked);
  };

  const handleTextChange = (text: String) => {
    setNewText(text);
  };

  const submitItem = () => {
    const text = newText.trim();
    if (text.length > 0) {
      onSubmit({ checked, text });
      setNewText(Strings.EmptyString);
      setChecked(Strings.Unchecked);
    }
  };

  return (
    <View style={[styles.itemContainer, { backgroundColor: theme.itemContainer }]}>
      <View style={styles.radioButtonContainer}>
        <RadioButton
          value={checked}
          onRadioButtonPress={handleRadioButtonPress}
          checkedColor={theme.backgroundRadioButton}
          borderColor={theme.borderRadioButton}
        />
      </View>
      <View style={styles.textInputContainer}>
        <TextInput
          ref={inputRef}
          style={
            checked == Strings.Checked
              ? [styles.textInput, { color: theme.itemTextChecked }, styles.linethrough]
              : [styles.textInput, { color: theme.itemNewText }]
          }
          editable
          maxLength={55}
          multiline={false}
          onChangeText={handleTextChange}
          placeholder={Strings.CreateTodo.toString()}
          placeholderTextColor={theme.itemPlaceholderNewText}
          autoFocus
          onSubmitEditing={submitItem}
        >
          {newText}
        </TextInput>
      </View>
    </View>
  );
};

export default CreateItem;

const styles = StyleSheet.create({
  itemContainer: {
    height: itemHeight,
    width: "85%",
    marginTop: "16%",
    borderRadius: 5,
    flexDirection: "row",
    justifyContent: "flex-start",
    alignItems: "center",
  },
  radioButtonContainer: {
    marginHorizontal: "5%",
  },
  textInputContainer: {
    flex: 0.9,
  },
  textInput: {
    fontFamily: "JosefinSans-Regular",
    fontSize: 16,
  },
  linethrough: {
    textDecorationLine: "line-through",
  },
});
