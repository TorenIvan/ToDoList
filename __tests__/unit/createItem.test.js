import React from "react";
import { TextInput } from "react-native";
import { create, act } from "react-test-renderer";
import CreateItem from "../../src/containers/CreateItem";

const tree = create(<CreateItem />);

describe("Create Item Container Testing", () => {
  test("snapshot", () => {
    expect(tree).toMatchSnapshot();
  });

  it("should render the placeholder when checkbox is not pressed no test is written", () => {
    expect(tree.root.findByType(TextInput).props.placeholder).toBe("Create a new todo");
  });

  it("should render the placeholder when checkbox is pressed and no text is written", () => {
    const checkbox = tree.root.findByProps({ testId: "CheckBox" }).props;
    act(() => checkbox.onRadioButtonPress());

    const textInput = tree.root.findByType(TextInput).props;
    expect(textInput.placeholder).toBe("Create a new todo");
  });

  it("should show the real-time text when textInput is written and checkbox is not pressed", () => {
    const textInput = tree.root.findByType(TextInput).props;
    act(() => textInput.onChangeText("Some Text"));

    const textInputUpdated = tree.root.findByType(TextInput).props;
    expect(textInputUpdated.children).toEqual("Some Text");
  });

  it("should show the real-time text with underline when textInput is written and checkbox is pressed", () => {});

  test("text input blurs when keyboard hides", () => {});
});
