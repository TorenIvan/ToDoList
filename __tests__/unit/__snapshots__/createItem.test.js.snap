// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Item Container Testing snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 5,
        "flexDirection": "row",
        "height": 106.72,
        "justifyContent": "flex-start",
        "marginTop": "16%",
        "width": "85%",
      },
      Object {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <View
    style={
      Object {
        "marginHorizontal": "5%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 110,
            "borderWidth": 2,
            "height": 25,
            "justifyContent": "center",
            "width": 25,
          },
          Object {
            "borderColor": "hsl(233, 11%, 84%)",
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 0.9,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoFocus={true}
      editable={true}
      maxLength={56}
      multiline={false}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Create a new todo"
      placeholderTextColor="#C9C8CD"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "fontFamily": "JosefinSans-Regular",
            "fontSize": 16,
          },
          Object {
            "color": "#88878F",
          },
        ]
      }
      underlineColorAndroid="transparent"
    >
      
    </TextInput>
  </View>
</View>
`;
